libraries/AP_NavEKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,678 @@
+	if (!ahrs_tilt_aligned) {
+		const float accel_norm_sq = ahrs_accel.length_squared();
+		const float upper_accel_limit = GRAVITY_MSS * 1.1f;
+		const float lower_accel_limit = GRAVITY_MSS * 0.9f;
+		const bool ok_to_align = ((accel_norm_sq > lower_accel_limit * lower_accel_limit &&
+			  accel_norm_sq < upper_accel_limit * upper_accel_limit));
+		if (ok_to_align) {
+			initialise();
+			alignTilt();
+			ahrs_tilt_aligned = true;
+		}
+		return;
+	}
+
+	// Calculate common variables used by the AHRS prediction models
+	ahrs_accel_norm = ahrs_accel.length();
+
+	// Calculate AHRS acceleration fusion gain using a quadratic weighting function that is unity at 1g
+	// and zero at the min and max g limits. This reduces the effect of large g transients on the attitude
+	// esitmates.
+	float EKFGSF_ahrs_ng = ahrs_accel_norm / GRAVITY_MSS;
+	if (EKFGSF_ahrs_ng > 1.0f) {
+		if (true_airspeed > FLT_EPSILON) {
+			// When flying in fixed wing mode we need to allow for more positive g due to coordinated turns
+			// Gain varies from unity at 1g to zero at 2g
+			accel_gain = EKFGSF_tiltGain * sq(2.0f - EKFGSF_ahrs_ng);
+		} else if (accel_gain <= 1.5f) {
+			// Gain varies from unity at 1g to zero at 1.5g
+			accel_gain = EKFGSF_tiltGain * sq(3.0f - 2.0f * EKFGSF_ahrs_ng);
+		} else {
+			// Gain is zero above max g
+			accel_gain = 0.0f;
+		}
+	} else if (accel_gain > 0.5f) {
+		// Gain varies from zero at 0.5g to unity at 1g
+		accel_gain = EKFGSF_tiltGain * sq(2.0f * EKFGSF_ahrs_ng - 1.0f);
+	} else {
+		// Gain is zero below min g
+		accel_gain = 0.0f;
+	}
+
+	// Always run the AHRS prediction cycle for each model
+	for (uint8_t mdl_idx = 0; mdl_idx < N_MODELS_EKFGSF; mdl_idx ++) {
+		predict(mdl_idx);
+	}
+
+	// The 3-state EKF models only run when flying to avoid corrupted estimates due to operator handling and GPS interference
+	if (vel_data_updated && run_ekf_gsf) {
+		if (!vel_fuse_running) {
+			// Perform in-flight alignment
+			for (uint8_t mdl_idx = 0; mdl_idx < N_MODELS_EKFGSF; mdl_idx ++) {
+				// Use the firstGPS  measurement to set the velocities and corresponding variances
+				EKF[mdl_idx].X[0] = vel_NE[0];
+				EKF[mdl_idx].X[1] = vel_NE[1];
+				EKF[mdl_idx].P[0][0] = sq(fmaxf(vel_accuracy, 0.5f));
+				EKF[mdl_idx].P[1][1] = EKF[mdl_idx].P[0][0];
+			}
+			alignYaw();
+			vel_fuse_running = true;
+		} else {
+			float total_w = 0.0f;
+			float newWeight[(uint8_t)N_MODELS_EKFGSF];
+			for (uint8_t mdl_idx = 0; mdl_idx < N_MODELS_EKFGSF; mdl_idx ++) {
+				// Update states and covariances using GPS NE velocity measurements fused as direct state observations
+				correct(mdl_idx);
+
+				// Calculate weighting for each model assuming a normal distribution
+				newWeight[mdl_idx]= fmaxf(gaussianDensity(mdl_idx) * GSF.weights[mdl_idx], 0.0f);
+				total_w += newWeight[mdl_idx];
+			}
+
+			// Normalise the sume of weights to unity
+			if (vel_fuse_running && total_w > 0.0f) {
+				float total_w_inv = 1.0f / total_w;
```

Comment:
testing in SITL I got a FPE on this line, total_w was 1.0e-41. Testing for > 0 isn't enough